### PR TITLE
Build Avocado SSO

### PR DESCRIPTION
**Pull request by Sardis93**
**Related Issue/Ticket:**

Related Issue: https://openy.atlassian.net/browse/PRODDEV-272

## Steps to test:

- [ ] Enable **Avocado Virtual YMCA integration** (machine name is `openy_gc_auth_avocado_sso`).
- [ ] Navigate to Virtual YMCA Auth settings page (found on path `admin/openy/virtual-ymca/gc-auth-settings` or in the navigation bar: Virtual Y -> Virtual YMCA settings -> GC auth settings).
- [ ] Make sure the **Avocado SSO provider** is enabled. Otherwise, click on “Active” checkbox within **Avocado SSO provider** row and save the form.
- [ ] Go to **Avocado SSO provider** configuration form: click on “Edit” button on the appropriate row.
- [ ] Provide values for the following fields: **Authentication server**, **Client Id**, **Client Secret**. Save the form afterwards.
- [ ] Go to a page with Virtual Y login paragraph, i.e. `/virtual-y-login`. Create one if it does not exist.
- [ ] Hit “Enter Virtual Y”.
- [ ] Fill **Username** & **Password** fields on the form that has appeared. Click “Log in to Sandbox”.
- [ ] Get redirected back to the page & confirm you are logged in.

## Quality checks:

Please check these boxes to confirm this PR covers the following cases:

- Maintaining our upgrade path is essential. Check one or the other:
  - [ ] This PR  provides updates via `hook_update_N` or other means.
  - [x] No updates are necessary for this change.
- Front end fixes should be tested against all of the Open Y Themes.
  - [ ] Tested against Carnation
  - [ ] Tested against Lily
  - [ ] Tested against Rose
  - [ ] This change does not contain front-end fixes.
- [x] I have flagged this PR "Needs Review" or pinged the VY devs/QA
  team in Slack

### CODE REVIEW

**Comment by anpolimus** [comment]
retest

**Comment by Sardis93** [comment]
retest

File: modules/openy_gc_auth/modules/openy_gc_auth_avocado_sso/config/install/openy_gc_auth.provider.avocado_sso.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+authentication_server: 'https://avocado-dev-6-developer-edition.eu30.force.com/'
```
Comment (anpolimus) [inline modules/openy_gc_auth/modules/openy_gc_auth_avocado_sso/config/install/openy_gc_auth.provider.avocado_sso.yml:1]:
```suggestion
authentication_server: 'https://<yourdomain>.force.com/'
```

File: modules/openy_gc_auth/modules/openy_gc_auth_avocado_sso/config/install/openy_gc_auth.provider.avocado_sso.yml [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+ymca_extension: 'ts_avo'
+client_id: ''
+client_secret: ''
+location_code: 'LOC-002427'
```
Comment (anpolimus) [inline modules/openy_gc_auth/modules/openy_gc_auth_avocado_sso/config/install/openy_gc_auth.provider.avocado_sso.yml:5]:
```suggestion
location_code: 'LOC-XXXXXX'
```

File: modules/openy_gc_auth/modules/openy_gc_auth_avocado_sso/src/Plugin/GCIdentityProvider/AvocadoSSO.php [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public function defaultConfiguration(): array {
+    return [
+      'authentication_server' => 'https://avocado-dev-6-developer-edition.eu30.force.com/',
+      'ymca_extension' => 'ts_avo',
+      'client_id' => '',
+      'client_secret' => '',
+      'location_code' => 'LOC-002427',
```
Comment (anpolimus) [inline modules/openy_gc_auth/modules/openy_gc_auth_avocado_sso/src/Plugin/GCIdentityProvider/AvocadoSSO.php:97]:
```suggestion
      'location_code' => 'LOC-XXXXXX',
```

File: modules/openy_gc_auth/modules/openy_gc_auth_avocado_sso/src/Plugin/GCIdentityProvider/AvocadoSSO.php [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+      $container->get('form_builder')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function defaultConfiguration(): array {
+    return [
+      'authentication_server' => 'https://avocado-dev-6-developer-edition.eu30.force.com/',
```
Comment (anpolimus) [inline modules/openy_gc_auth/modules/openy_gc_auth_avocado_sso/src/Plugin/GCIdentityProvider/AvocadoSSO.php:93]:
```suggestion
      'authentication_server' => 'https://<yourdomain>.force.com/',
```

File: modules/openy_gc_auth/modules/openy_gc_auth_avocado_sso/src/SSOClient.php [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+          $this->configAvocadoSSO->get('ymca_extension'),
+          self::ENDPOINT_COMMUNITY_EVENT_LOG
+        );
+
+      $response = $this->httpClient->request(
+        'POST',
+        $request_uri,
+        [
+          'headers' => [
+            'Authorization' => "Bearer " . $this->getAccessToken(),
```
Comment (anpolimus) [inline modules/openy_gc_auth/modules/openy_gc_auth_avocado_sso/src/SSOClient.php:302]:
```suggestion
            'Authorization' => 'Bearer ' . $this->getAccessToken(),
```

File: modules/openy_gc_auth/modules/openy_gc_auth_avocado_sso/src/SSOClient.php [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+          'form_params' => [
+            'memberBarcode' => $membership_barcode,
+            'locationCode' => $this->configAvocadoSSO->get('location_code'),
+          ],
+        ]);
+
+      $body = json_decode((string) $response->getBody(), FALSE);
+      if (
+        !isset($body->data)
+        || $body->data->eventLogMessage !== "Success"
```
Comment (anpolimus) [inline modules/openy_gc_auth/modules/openy_gc_auth_avocado_sso/src/SSOClient.php:313]:
```suggestion
        || $body->data->eventLogMessage !== 'Success'
```

**Review by anpolimus — APPROVED** [review]

**Pull request merged**